extendr-macros/src/wrappers.rs

```diff
@@ -102,17 +100,48 @@ pub fn make_function_wrappers(
     //     }
     // }
     // ```
+    //
     wrappers.push(parse_quote!(
         #[no_mangle]
         #[allow(non_snake_case, clippy::not_unsafe_ptr_arg_deref)]
         pub extern "C" fn #wrap_name(#formal_args) -> extendr_api::SEXP {
-            unsafe {
-                use extendr_api::robj::*;
+            use extendr_api::robj::*;
+            let wrap_result_state: std::result::Result<
+                std::result::Result<Robj, extendr_api::Error>,
+                Box<dyn std::any::Any + Send>
+            > = unsafe {
                 #( #convert_args )*
-                extendr_api::handle_panic(#panic_str, ||
-                    extendr_api::Robj::from(#call_name(#actual_args)).get()
-                )
+                std::panic::catch_unwind(||-> std::result::Result<Robj, extendr_api::Error> {
+                    Ok(extendr_api::Robj::from(#call_name(#actual_args)))
+                })
+            };
+            // any obj created in above unsafe scope, which are not moved into wrap_result_state are now dropped
+
+
+            match wrap_result_state {
+                Ok(Ok(zz)) => {
+                    return unsafe { zz.get() };
+                }
+                // any conversion error bubbled from #actual_args conversions of incomming args from R.
+                Ok(Err(conversion_err)) => {
+                    let err_string = conversion_err.to_string();
+                    drop(conversion_err); // try_from=true errors contain Robj, this must be dropped to not leak
```

Comment:
I may be missing the context, but why is it not dropped at the end of the scope (as all Rust things are)? Is it because `::throw_r_error()` breaks things up?

Reply:
Paraphrasing the writings of @yutannihilation . Rust has a stack and perhaps some heap allocations. Performing longjump will not unwind the stack and free heap allocations. But also panic unwind is discouraged in rust docs and stated as UB when returning to another language as R.

Previous attempts to manually drop before throw_r_error() did not fully succeed because they only dropped one or some scopes on top of the stack. By first returning `Result` the whole conversion stack can be dropped, and then only a single object needs to be dealt with. That is a saner way to deal with all edge cases and it emperically solves any known case of leaking out of ~150 test cases.

I have used with r-polars and there is none of 1200 unit tests which fail.